### PR TITLE
[SPARK-30588][DOC] Document CLUSTER BY Clause of SELECT statement in SQL Reference.

### PR DESCRIPTION
**Pull request by dilipbiswal**
### What changes were proposed in this pull request?
Document CLUSTER BY clause of SELECT statement in SQL Reference Guide.

### Why are the changes needed?
Currently Spark lacks documentation on the supported SQL constructs causing
confusion among users who sometimes have to look at the code to understand the
usage. This is aimed at addressing this issue.

### Does this PR introduce any user-facing change?
Yes. 

**Before:**
There was no documentation for this.

**After.**
Screen Shot 2020-01-20 at 2 59 05 PM
Screen Shot 2020-01-20 at 2 59 19 PM

### How was this patch tested?
Tested using jykyll build --serve

### CODE REVIEW

File: docs/sql-ref-syntax-qry-select-clusterby.md [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+ 
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+---
+The <code>CLUSTER BY</code> clause is used to first repartition the data based
+on the input expressions and then sort the data within each partition. This is
+is semantically equivalent to performing <code>DISTRIBUTE BY</code> followed by
```
Comment (huaxingao) [inline docs/sql-ref-syntax-qry-select-clusterby.md]: ```This is is``` -> ```This is```

File: docs/sql-ref-syntax-qry-select-clusterby.md [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+<code>SORT BY</code>. This cause only ensures that the resultant rows are
```
Comment (huaxingao) [inline docs/sql-ref-syntax-qry-select-clusterby.md]: Link to ```DISTRIBUTE BY``` and ```SORT BY```?

File: docs/sql-ref-syntax-qry-select-clusterby.md [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+sorted within each partition and does not guarantee total ordering of output,
```
Comment (huaxingao) [inline docs/sql-ref-syntax-qry-select-clusterby.md]: You accidentally put ```,``` in the end of the sentence.

File: docs/sql-ref-syntax-qry-select-clusterby.md [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+{% highlight sql %}
+CREATE TABLE person (name STRING, age INT);
+INSERT INTO person VALUES ('Zen Hui', 25), 
+                          ('Anil B', 18), 
+                          ('Shone S', 16), 
+                          ('Mike A', 25),
+                          ('John A', 18), 
+                          ('Jack N', 16);
+-- Reduce the number of shuffle partitions to 2 to illustrate the behaviour of `CLUSTER BY`.
+-- Its easier to see the clustering and sorting behaviour with less number of partitions.
```
Comment (huaxingao) [inline docs/sql-ref-syntax-qry-select-clusterby.md]: ```Its``` -> ```It's```?

File: docs/sql-ref-syntax-qry-select-clusterby.md [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+SET spark.sql.shuffle.partitions = 2;
+                        
+-- Select the rows with no ordering. Please note that without any sort directive, the results
+-- of the query is not deterministic. Its included here to show the difference in behaviour
```
Comment (huaxingao) [inline docs/sql-ref-syntax-qry-select-clusterby.md]: ```results of the query is``` -> ```result of the query is```???
```Its``` -> ```It's```?

File: docs/sql-ref-syntax-qry-select-clusterby.md (hunk 2 above)
Comment (dilipbiswal) [reply docs/sql-ref-syntax-qry-select-clusterby.md]: @huaxingao Lets do in the finalization pr when these links are available.

File: docs/sql-ref-syntax-qry-select-clusterby.md [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+                          ('Mike A', 25),
+                          ('John A', 18), 
+                          ('Jack N', 16);
+-- Reduce the number of shuffle partitions to 2 to illustrate the behaviour of `CLUSTER BY`.
+-- It's easier to see the clustering and sorting behaviour with less number of partitions.
+SET spark.sql.shuffle.partitions = 2;
+                        
+-- Select the rows with no ordering. Please note that without any sort directive, the results
+-- of the query is not deterministic. It's included here to show the difference in behaviour 
+-- of a query when `CLUSTER BY` is not used vs when its used. The query below produces rows
```
Comment (huaxingao) [inline docs/sql-ref-syntax-qry-select-clusterby.md]: ```when its used``` -> ```when it's used```
Sorry, I didn't catch this in the first review.

File: docs/sql-ref-syntax-qry-select-clusterby.md [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+---
+The <code>CLUSTER BY</code> clause is used to first repartition the data based
+on the input expressions and then sort the data within each partition. This is
+semantically equivalent to performing <code>DISTRIBUTE BY</code> followed by
+<code>SORT BY</code>. This cause only ensures that the resultant rows are
+sorted within each partition and does not guarantee total ordering of output.
```
Comment (maropu) [inline docs/sql-ref-syntax-qry-select-clusterby.md]: `total ordering of output.` -> `a total order of output.`

File: docs/sql-ref-syntax-qry-select-clusterby.md [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+  <dt><code><em>expression</em></code></dt>
+  <dd>
+    Specifies combination of one or more values, operators and SQL functions that results in a value.
+  </dd>
+</dl>
+
+### Examples
+{% highlight sql %}
+CREATE TABLE person (name STRING, age INT);
+INSERT INTO person VALUES ('Zen Hui', 25),
```
Comment (maropu) [inline docs/sql-ref-syntax-qry-select-clusterby.md]: Do you have any reference for this format? I checked the PgSQL document and they have the other form like this;
```
INSERT INTO person VALUES
    ('Zen Hui', 25), 
    ('Anil B', 18), 
    ('Shone S', 16), 
    ('Mike A', 25),
    ('John A', 18), 
    ('Jack N', 16);
```
https://www.postgresql.org/docs/current/sql-insert.html#id-1.9.3.152.9

Comment (dilipbiswal) [reply docs/sql-ref-syntax-qry-select-clusterby.md]: @maropu No.. i don't :-) Lets me fix this for all the prs thats open currently. I will fix the one that are closed in the finalization pr.

File: docs/sql-ref-syntax-qry-select-clusterby.md (hunk 6 above)
Comment (dilipbiswal) [reply docs/sql-ref-syntax-qry-select-clusterby.md]: @huaxingao thanks.

**Comment by srowen** [comment]
Merged to master

**Pull request closed without merging**